nixos/modules/services/web-apps/immich.nix

```diff
@@ -0,0 +1,297 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.services.immich;
+  isPostgresUnixSocket = lib.hasPrefix "/" cfg.database.host;
+  postgresEnv =
+    if isPostgresUnixSocket then
+      { DB_URL = "socket://${cfg.database.host}?dbname=${cfg.database.name}"; }
+    else
+      {
+        DB_HOSTNAME = cfg.database.host;
+        DB_PORT = toString cfg.database.port;
+        DB_DATABASE_NAME = cfg.database.name;
+        DB_USERNAME = cfg.database.user;
+      };
+  isRedisUnixSocket = lib.hasPrefix "/" cfg.redis.host;
+  redisEnv =
+    if isRedisUnixSocket then
+      { REDIS_SOCKET = "unix:${cfg.redis.host}"; }
+    else
+      {
+        REDIS_PORT = toString cfg.redis.port;
+        REDIS_HOSTNAME = cfg.redis.host;
+      };
+
+  commonServiceConfig = {
+    Type = "simple";
+    Restart = "on-failure";
+    RestartSec = 3;
+
+    # Hardening
+    CapabilityBoundingSet = "";
+    NoNewPrivileges = true;
+    PrivateUsers = true;
+    PrivateTmp = true;
+    PrivateDevices = true;
+    PrivateMounts = true;
+    ProtectClock = true;
+    ProtectControlGroups = true;
+    ProtectHome = true;
+    ProtectHostname = true;
+    ProtectKernelLogs = true;
+    ProtectKernelModules = true;
+    ProtectKernelTunables = true;
+    RestrictAddressFamilies = [
+      "AF_INET"
+      "AF_INET6"
+      "AF_UNIX"
+    ];
+    RestrictNamespaces = true;
+    RestrictRealtime = true;
+    RestrictSUIDSGID = true;
+  };
+  inherit (lib)
+    types
+    mkIf
+    mkOption
+    mkEnableOption
+    ;
+in
+{
+  options.services.immich = {
+    enable = mkEnableOption "Immich";
+    package = lib.mkPackageOption pkgs "immich" { };
+
+    mediaLocation = mkOption {
+      type = types.path;
+      default = "/var/lib/immich";
+      description = "Directory used to store media files. If it is not the default, the directory has to be created manually such that the immich user is able to read and write to it.";
+    };
+    environment = mkOption {
+      type = types.submodule { freeformType = types.attrsOf types.str; };
+      default = { };
+      example = {
+        IMMICH_LOG_LEVEL = "verbose";
+      };
+      description = ''
+        Extra configuration environment variables. Refer to the [documentation](https://immich.app/docs/install/environment-variables) for options tagged with 'server', 'api' or 'microservices'.
+      '';
+    };
+    secretsFile = mkOption {
+      type = types.nullOr (
+        types.str
+        // {
+          # We don't want users to be able to pass a path literal here but
+          # it should look like a path.
+          check = it: lib.isString it && lib.types.path.check it;
```

Comment:
I don't think these homegrown solutions are a good idea until something like https://github.com/NixOS/nixpkgs/pull/78640 is resolved.

Reply:
This was done in response to another review comment by @Atemu I think

Reply:
Until https://github.com/NixOS/nixpkgs/pull/78640 is resolved, "home-grown solutions" are the only option.

Reply:
Then we should make of them official immediately and improve upon it later.